dev/AppNotifications/ShellLocalization.cpp

```diff
@@ -0,0 +1,272 @@
+
+    ULARGE_INTEGER lnbuffer{ INT_MAX };
+    THROW_IF_FAILED(spImageStream->CopyTo(pStream.get(), lnbuffer, nullptr /* pointer to number of bytes read */, nullptr /* pointer to number of bytes written */));
+}
+
+void WriteHIconToPngFile(wil::unique_hicon const& hIcon, _In_ PCWSTR pszFileName)
+{
+    auto wicImagingFactory{ winrt::create_instance<IWICImagingFactory>(CLSID_WICImagingFactory, CLSCTX_INPROC_SERVER) };
+
+    winrt::com_ptr<IWICBitmap> wicBitmap;
+    THROW_IF_FAILED(wicImagingFactory->CreateBitmapFromHICON(hIcon.get(), wicBitmap.put()));
+
+    // Create stream to save the HICON to.
+    winrt::com_ptr<IStream> spStream;
+    THROW_IF_FAILED(CreateStreamOnHGlobal(nullptr /* handle */, TRUE, spStream.put()));
+
+    winrt::com_ptr<IWICBitmapSource> wicBitmapSource;
+    wicBitmapSource = wicBitmap;
+
+    SaveImageWithWIC(wicImagingFactory, wicBitmapSource, GUID_ContainerFormatPng, spStream);
+
+    // Write out the stream to a file.
+    winrt::com_ptr<IStream> spStreamOut;
+    THROW_IF_FAILED(SHCreateStreamOnFileEx(pszFileName, STGM_CREATE | STGM_WRITE | STGM_SHARE_DENY_WRITE, FILE_ATTRIBUTE_NORMAL, TRUE, nullptr, spStreamOut.put()));
+
+    STATSTG statstg{};
+    THROW_IF_FAILED(spStream->Stat(&statstg, STATFLAG_NONAME));
+
+    THROW_IF_FAILED(IStream_Reset(spStream.get()));
+
+    THROW_IF_FAILED(spStreamOut->SetSize(statstg.cbSize));
+
+    THROW_IF_FAILED(spStream->CopyTo(spStreamOut.get(), statstg.cbSize, nullptr /* pointer to number of bytes read */, nullptr /* pointer to number of bytes written */));
+
+    THROW_IF_FAILED(spStreamOut->Commit(STGC_DANGEROUSLYCOMMITMERELYTODISKCACHE));
+}
+
+HRESULT IsIconFileExtensionSupported(std::filesystem::path const& iconFilePath)
+{
+    const auto fileExtension = iconFilePath.extension();
+
+    std::string lowercaseFileExtension{ fileExtension.u8string() };
+
+    std::transform(lowercaseFileExtension.begin(), lowercaseFileExtension.end(), lowercaseFileExtension.begin(),
+        [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
+
+    const bool isFileExtensionSupported =
+        lowercaseFileExtension == ".ico" || lowercaseFileExtension == ".bmp" || lowercaseFileExtension == ".jpg" || lowercaseFileExtension == ".png";
+
+    THROW_HR_IF(E_UNEXPECTED, isFileExtensionSupported);
+    return S_OK;
+}
+
+inline std::wstring RetrieveLocalFolderPath()
+{
+    wil::unique_cotaskmem_string localAppDataPath;
+    THROW_IF_FAILED(SHGetKnownFolderPath(FOLDERID_LocalAppData, 0 /* flags */, nullptr /* access token handle */, &localAppDataPath));
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK
+    std::path localFolderPath{ std::wstring(localAppDataPath.get()) +
+        Microsoft::Windows::AppNotifications::Helpers::c_localMicrosoftFolder +
+        Microsoft::Windows::AppNotifications::Helpers::c_localWindowsAppSDKFolder };
+
+    if (!std::exists(localFolderPath))
```

Comment:
I'm curious why we need to create a directory at all here. IF the file is saved, wouldn't the directory automatically be created? So this looks like overhead.

Reply:
@sharath2727 ?

Reply:
The API WriteHICONtoPng when it takes in a filepath as an argument doesn't create windowsappsdk folder (if it doesn't exist already) and throws if the path doesn't exist. So that is why we have this overhead of making sure the windowsappsdk folder exists in the local path.